### PR TITLE
feat(eslint-plugin)!: [array-type] add "default", "readonly" options

### LINKED ISSUE #635: [array-type] add `ReadonlyArray` option
**Repro**

```TS
// "typescript/array-type": ["error", "array"]
const x: Array<number> = [1];
// fixes to
const x: number[] = [1];
```

```TS
// "typescript/array-type": ["error", "ReadonlyArray"]
const x: ReadonlyArray<number> = [1];
// fixes to
const x: readonly number[] = [1];
```

```TS
// "typescript/array-type": ["error", "array", "ReadonlyArray"]
const x: ReadonlyArray<number> = [1];
const y: Array<number> = [1];
// fixes to
const x: readonly number[] = [1];
const y: number[] = [1];
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.10.2` |
| `@typescript-eslint/parser`        | `1.10.2` |
| `TypeScript`                       | `3.5.2` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `8.15.0` |
| `npm`                              | `6.4.1` |

### PR DESCRIPTION
**Pull request by a-tarasyuk**
BREAKING CHANGE: changes config structure

Fixes #635

### CODE REVIEW

**Comment by a-tarasyuk** [comment]
@bradzacher ok, will change

**Comment by a-tarasyuk** [comment]
@bradzacher 

```TS
type Options = [
  {
    // default case for all arrays
    default: ArrayOption,
    // optional override for readonly arrays
    readonly?: ArrayOption,
  },
];
```

The new option definition should behave like so

```TS
// "array-type": ["error", { "default": "array" }]
const x: ReadonlyArray<number> = [1];
const x: Array<number> = [1];
// fixes to
const x: ReadonlyArray<number> = [1];
const x: number[] = [1];
```

```TS
// "array-type": ["error", { "default": "array", "readonly": "array" }]
const x: ReadonlyArray<number> = [1];
const y: Array<number> = [1];
// fixes to
const x: readonly number[] = [1];
const y: number[] = [1];
```

```TS
// "array-type": ["error", { "default": "generic" }]
const a: string[] = [];
const x: readonly number[][] = [];
// fixes to
const a: Array<string> = [];
const x: readonly number[][] = [];
```

```TS
// "array-type": ["error", { "default": "generic", "readonly": "generic" }]
const a: string[] = [];
const x: readonly number[][] = [];
// fixes to
const a: Array<string> = [];
const x: ReadonlyArray<Array<number>> = [];
```
Is it right?

**Comment by bradzacher** [comment]
Almost!

Sorry, I didn't explain my intention clearly enough:
The default config should be this `{default:'array', readonly: options.default}`, i.e. the `readonly` option mirrors the `default` option unless explicitly provided.

So your examples were almost right.
The new option definition should behave like so

```diff
 // "array-type": ["error", { "default": "array" }]
 const x: ReadonlyArray<number> = [1];
 const x: Array<number> = [1];
 // fixes to
-const x: ReadonlyArray<number> = [1];
+const x: readonly number[] = [1];
 const x: number[] = [1];
```

```TS
// "array-type": ["error", { "default": "array", "readonly": "array" }]
const x: ReadonlyArray<number> = [1];
const y: Array<number> = [1];
// fixes to
const x: readonly number[] = [1];
const y: number[] = [1];
```

```diff
 // "array-type": ["error", { "default": "generic" }]
 const a: string[] = [];
 const x: readonly number[][] = [];
 // fixes to
 const a: Array<string> = [];
-const x: readonly number[][] = [];
+const x: ReadonlyArray<Array<number>> = [];
```

```TS
// "array-type": ["error", { "default": "generic", "readonly": "generic" }]
const a: string[] = [];
const x: readonly number[][] = [];
// fixes to
const a: Array<string> = [];
const x: ReadonlyArray<Array<number>> = [];
```

And of course the situations where the two differ:
```TS
// "array-type": ["error", { "default": "array", "readonly": "generic" }]
const a: string[] = [];
const x: readonly number[][] = [];
// fixes to
const a: string[] = [];
const x: ReadonlyArray<number[]> = [];
```
```TS
// "array-type": ["error", { "default": "generic", "readonly": "array" }]
const a: string[] = [];
const x: readonly number[][] = [];
// fixes to
const a: Array<string> = [];
const x: readonly Array<number>[] = [];
```

**Comment by a-tarasyuk** [comment]
@bradzacher Which option can skip check for `ReadonlyArray`?

**Comment by bradzacher** [comment]
I'm not sure I quite understand - do you mean you were looking for an option to ignore validating readonly arrays completely? I don't think that's a valid state for this rule, as you then can have inconsistent definitions.

`{ "default": "array" }` will enforce that all your array types use `(readonly)? foo[]`.
`{ "default": "generic" }` will enforce that all your arrays use `(Readonly)?Array<foo>`.
`{ "default": "array", "readonly": "generic" }` will enforce that all mutable arrays use `foo[]` and your readonly arrays use `ReadonlyArray<foo>`.
`{ "default": "generic", "readonly": "array" }` will enforce that all mutable arrays use `Array<foo>` and your readonly arrays use `readonly foo[]`.

The `"array-simple"` option will interop similarly, but I cbf enumerating those examples :)

**Comment by a-tarasyuk** [comment]
> Do you mean you were looking for an option to ignore validating readonly arrays completely? I don't think that's a valid state for this rule, as you then can have inconsistent definitions.

It was the main idea to add an option to "array", to have the opportunity to ignore `ReadonlyArray`.

**Comment by bradzacher** [comment]
I think you've misunderstood me.
`{ "default": "array", "readonly": "generic" }` will enforce that all mutable arrays use `foo[`] and your readonly arrays use `ReadonlyArray<foo>`.

Which is almost exactly the same as tslints `array` setting.

The difference being that tslint will let you use `readonly foo[]`, where as this configuration will enforce that you instead use `ReadonlyArray<foo>`

File: packages/eslint-plugin/tests/rules/array-type.test.ts [hunk 1]
```diff
@@ -10,175 +10,201 @@ ruleTester.run('array-type', rule, {
...
+    {
+      code: 'let a: ReadonlyArray<number[]> = [[]]',
+      options: [{ default: 'array', readonly: 'generic' }],
+    },
+    {
+      code: `let a: Array<string> = []`,
+      options: [{ default: 'generic', readonly: 'array' }],
+    },
+    {
+      code: `let a: readonly Array<number> = []`,
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/rules/array-type.test.ts]: this is not valid typescript
```suggestion
      code: `let a: readonly number[] = []`,
```

File: packages/eslint-plugin/tests/rules/array-type.test.ts [hunk 2]
```diff
@@ -10,175 +10,201 @@ ruleTester.run('array-type', rule, {
...
+      options: [{ default: 'generic', readonly: 'array' }],
+    },
+    {
+      code: `let a: readonly Array<Array<number>> = [[]]`,
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/rules/array-type.test.ts]: ditto
```suggestion
      code: `let a: readonly Array<number>[] = [[]]`,
```

File: packages/eslint-plugin/tests/rules/array-type.test.ts (hunk 1 above)
Comment (a-tarasyuk) [reply packages/eslint-plugin/tests/rules/array-type.test.ts]: @bradzacher Oops! Will fix. Thanks. :+1:

**Comment by octogonz** [comment]
@bradzacher I noticed that the default for this rule is to require `readonly string[]` instead of `ReadonlyArray<string>`.  This is a bleeding edge syntax that produces a syntax error for TypeScript 3.4 or earlier.  When I enabled this rule and ran the fixer, we received reports that our NPM package was no longer usable with older releases of TypeScript.

By contrast, requiring `string[]` instead of `Array<string>` seems idiomatic and perfectly fine.  Therefore, I wonder if the default should be like this instead:

```ts
"@typescript-eslint/array-type": [
  "error",
  {
    default: "array",
    readonly: "generic"
  }
],
```

What do you think?

**Comment by bradzacher** [comment]
I don't have a problem with recommending the bleeding edge, as the behaviour is documented in the readme.

I think it's better to recommend the most consistent syntax, and have people configure the support for their environment if they don't have the new syntax. Yours is the first complaint about it, so it might be uncommon to not be on 3.4? I don't know exactly.

Regardless, changing the default option is a breaking change, which means it'd have to wait for a while if we did want to make the change.

**Comment by octogonz** [comment]
If we measure number of projects, it's probably very uncommon to not be using the latest bleeding edge. Statistically the vast majority of projects are tiny and probably just got started. Whereas if we measure number of dollars earned, older compilers are probably way more likely: Upgrading TypeScript can require nontrivial effort for a production code base. Besides all the files that need fixes, it may also require upgrading `@types` packages, which may require upgrading the underlying library. I worked on a project one time where two people worked for two full weeks to get everything to compile again, and it wasn't even a major version increment for the compiler.

**Pull request merged**